### DISCUSSION TITLE
How to call a function when 2 objects stop overlapping

### DISCUSSION THREAD

**Discussion by NikRetaNCAM** [HaxeFlixel/flixel#2517] [Q&A] [answered]
So, for example, i have a constantly moving object (let me call it mov). When it crosses another object (let me name it stat), i want function a to be called, when mov and stat stop overlapping (mov goes beyond the stat's trigger) i want function b to be called

can someone please help?

**Comment by NikRetaNCAM** [comment]
or simply, when they overlap, some bool gets true, when they stop overlapping, bool gets false

**Comment by RichardBray** [comment, marked as answer]
So there are two ways to achieve what you want to do. I say two there are probably way more but only two that I know of. 
1. With Flixel's basic collision and overlapping
2. With [echo's](https://austineast.dev/echo/) more advanced collision and overlapping 

### 1. WIth Flixel

You can use [FlxG.overlap](https://api.haxeflixel.com/flixel/FlxG.html#overlap) to check if two objects have overlapped.
```haxe
FlxG.overlap(mov, stat, (_, _) -> functionA());
```

To detect if the two objects have exited overlapping with just plain old HaxeFlixel is a little more difficult.

What you could do is have a boolean variable that
 triggers when they overlap and in the update function check if they are not overlapping when that boolean variable is true.
 
 ```haxe
// just below class name
var objectsTouched = false;

// inside update() function
FlxG.overlap(mov, stat, (_, _) -> {
  functionA();
  objectsTouched = true;
});

if (!FlxG.overlap(mov, stat) && objectsTouched) {
  functionB();
}
 ```

_Note: The above code is not refactored so I understand that improvements can definitely be made to it_

<br/>

### 2. With Echo

This option has the caveat of first installing [echo-flixel](https://github.com/AustinEast/echo-flixel), putting it into your project.xml file, putting `using echo.FlxEcho;` at the top of your file along with the other imports and initializing echo (all not that complicated).

```haxe
// inside create() function
mov.add_body();
stat.add_body({mass: 0});

mov.listen(stat, {
  separate: false,
  enter: (_, _, _) -> functionA(),
  exit: (_, _) -> functionB(),
});
```

The choice is completely up to you.

**Reply by NikRetaNCAM** [reply]
thank you

sorry if my questions are dumb sometimes😅

**Reply by Geokureli** [reply]
@NicklGames this is a great question, and thanks for the fantastic answer @RichardBray !
